{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8491915-084d-439e-828d-c52cd231c4b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#lrRNAseq GAST (Global Alignment Search Tool) visualiser\n",
    "#run this notebook sequentially whilst filling in required variable paths and parameters\n",
    "#written in all capital letters e.g. TRANSCRIPT_TO_GENOME_BLAST_PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21090156-c74f-4e86-a0bf-10b3cbbe85b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#must have the following libraries installed and imported: Pandas, Numpy, Matplotlib, regex, os, Biopython, ast, operator, warnings.\n",
    "#must also have seqtk installed via bioconda i.e. conda install bioconda::seqtk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import os\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib\n",
    "from Bio import SeqIO\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from ast import literal_eval\n",
    "from operator import itemgetter\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "%run \"./bioinformatic_functions.ipynb\" #import functions from the \"bioinformatic_functions\" file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "facd25da-f884-44f4-91cd-97645bf8906b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Align long read RNAseq transcripts to a genomic region using BLAST with alignment parameters (-word_size, -evalue) of your choice.\n",
    "\n",
    "#import lrRNAseq BLAST alignments to genomic region\n",
    "TRANSCRIPT_TO_GENOME_BLAST_PATH = \"\"\n",
    "#e.g. TRANSCRIPT_TO_GENOME_BLAST_PATH = \"./lrRNAseq_to_Ef318A10.out\"\n",
    "all_blast = read_blast(TRANSCRIPT_TO_GENOME_BLAST_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23c0cb16-b41d-4b2f-a8c0-8b0e87db6ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate transcript filtration variables i.e. transcript coverage and maximal intron/alignment gap length\n",
    "grouped_blast = all_blast.groupby(\"qseqid\")\n",
    "\n",
    "all_blast[\"Tr_coverage\"] = 0\n",
    "cov_dict = {}\n",
    "dist_dict = {}\n",
    "for tr, temp_df in grouped_blast:\n",
    "    qlen = temp_df[\"qlen\"].max()\n",
    "    temp_df = temp_df.sort_values(by=\"qstart\").reset_index(drop=True)\n",
    "    q_overlaps, q_cov = calc_overlaps(temp_df, \"q\") \n",
    "    q_overlaps[\"Length\"] = q_overlaps[\"Right\"] - q_overlaps[\"Left\"] \n",
    "    coverage = 100 / qlen * q_overlaps[\"Length\"].sum()  \n",
    "    cov_dict[tr] = coverage\n",
    "\n",
    "    temp_df = temp_df.sort_values(by=\"sstart\").reset_index(drop=True)\n",
    "    q_overlaps, q_cov = calc_overlaps(temp_df, \"s\") \n",
    "    right = q_overlaps[\"Right\"].tolist()\n",
    "    left = q_overlaps[\"Left\"].tolist()\n",
    "    distances = []\n",
    "    if len(right)>1:\n",
    "        for i in range(len(right)-1):\n",
    "            distances.append(left[i+1]-right[i])\n",
    "    else:\n",
    "        distances = [0]\n",
    "    dist_dict[tr] = distances\n",
    "\n",
    "# all_blast.loc[all_blast[\"qseqid\"] == tr, \"Tr_coverage\"] = coverage\n",
    "all_blast[\"Tr_coverage\"] = all_blast[\"qseqid\"].apply(lambda x: cov_dict[x])\n",
    "all_blast[\"s_distances\"] = all_blast[\"qseqid\"].apply(lambda x: dist_dict[x])\n",
    "all_blast[\"max_distance\"] = all_blast[\"s_distances\"].apply(lambda x: max(x))\n",
    "all_blast[\"srange\"] = all_blast.apply(lambda x: [x.sstart, x.send], axis = 1)\n",
    "all_blast[\"qrange\"] = all_blast.apply(lambda x: [x.qstart, x.qend], axis = 1)\n",
    "all_blast[\"counts\"] = all_blast.groupby(\"qseqid\")[\"sseqid\"].transform(\"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cab593a7-2806-45b7-b69c-6260b3a2fa89",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov50=all_blast.sort_values(\"sstart\").reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbded456-a51c-43be-aa41-1f7278e0ff1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter by maximal intron length\n",
    "MAX_INTRON_LENGTH = 20000 #bp\n",
    "cov50 = cov50[cov50[\"max_distance\"]<MAX_INTRON_LENGTH]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "639d1877-baa6-4e00-bcbb-ba9dd93e3328",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter by minimal percentage of transcript length covered by the genome as per collective BLAST alignments\n",
    "MIN_COVERAGE = 25 #percent\n",
    "cov50 = cov50[cov50[\"Tr_coverage\"]>MIN_COVERAGE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c9432b-1f5c-4895-bf9e-c294292d63c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save filtered transcript ids to list in main directory\n",
    "FILTERED_TRANSCRIPT_IDS = \"\"\n",
    "#e.g. FILTERED_TRANCRIPT_IDS = \"./aligned_lrRNAseq_cov25_filtered.txt\"\n",
    "write_list(cov50.drop_duplicates(\"qseqid\")[\"qseqid\"].tolist(), FILTERED_TRANCRIPT_IDS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58eddf0d-6293-4700-9d00-9f25c36d6655",
   "metadata": {},
   "outputs": [],
   "source": [
    "#produce a FASTA file including only the transcripts that passed the filtration procedure\n",
    "#NOTE: must have seqtk to run #conda install bioconda::seqtk\n",
    "ALL_TRANSCRIPTS_PATH = \"\" # enter path to the bulk RNAseq file\n",
    "#e.g. ALL_TRANSCRIPTS_PATH = \"./all_transcripts.fasta\"\n",
    "FILTERED_TRANSCRIPTS_FASTA_PATH = \"\" #filtered transcripts FASTA path\n",
    "#e.g. FILTERED_TRANSCRIPTS_FASTA_PATH = \"./filtered_transcripts.fasta\"\n",
    "os.system(f\"seqtk subseq {ALL_TRANSCRIPTS_PATH} {FILTERED_TRANSCRIPT_IDS} > {FILTERED_TRANSCRIPTS_FASTA_PATH})\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "32332bfc-2bdf-4671-a94c-c6dbfaec1c98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.251478910446167\n"
     ]
    }
   ],
   "source": [
    "#import filtered transcripts FASTA and find ORF regions/stats\n",
    "BARCODE_LEN = 0 #trim n barcode characters from both ends of transcripts #set to 39 if working with data provided in the example\n",
    "cov50_fasta = parse_fasta(FILTERED_TRANSCRIPTS_FASTA_PATH,rm_barcode=BARCODE_LEN)[0]\n",
    "all_stats = find_tr_ORF_stats(cov50_fasta[\"seq\"].tolist(),cov50_fasta[\"id\"].tolist(),tissue=True, include_seq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f210200f-109d-432c-afd1-17ecb1c5f5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import FASTA of the genomic region of interest e.g. BAC\n",
    "GENOME_FASTA_PATH = \"\"\n",
    "#e.g. GENOME_FASTA_PATH = \"./Ef318A10.fst\"\n",
    "my_seq = \"\".join([line.strip() for line in open(GENOME_FASTA_PATH) if not line.startswith(\">\")])\n",
    "seq_len = len(my_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10987823-d2ce-4736-b059-564d0d07465e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate ORF overlap statistics\n",
    "all_s = all_stats[[\"id\", \"ORF_coords\", \"all_ORF_coords\", \"ORF_lengths\", \"num_ORFs_found\"]]\n",
    "all_s.rename(columns={\"id\":\"qseqid\"}, inplace=True)\n",
    "\n",
    "cov50_ranges = cov50.groupby(\"qseqid\").agg({\"srange\": \"sum\", \"qrange\": \"sum\"})\n",
    "cov50_ranges = cov50_ranges.reset_index()\n",
    "\n",
    "cov50_ranges[\"start\"] = cov50_ranges[\"srange\"].apply(lambda x: min(x))\n",
    "cov50_ranges[\"end\"] = cov50_ranges[\"srange\"].apply(lambda x: max(x))\n",
    "\n",
    "cov50_ranges[\"qstarts\"] = cov50_ranges[\"qrange\"].apply(lambda x: x[::2])\n",
    "cov50_ranges[\"TrRange\"] = cov50_ranges.apply(lambda x: [x[\"start\"], x[\"end\"]], axis=1)\n",
    "cov50_ranges[\"TrLen\"] = cov50_ranges[\"end\"] - cov50_ranges[\"start\"]\n",
    "\n",
    "cov50_ranges = cov50_ranges.merge(right=all_s, on=\"qseqid\",how=\"inner\")\n",
    "cov50_ranges[\"left_ORF_coords\"] = cov50_ranges[\"all_ORF_coords\"].apply(lambda x: [i[0] for i in x])\n",
    "cov50_ranges[\"right_ORF_coords\"] = cov50_ranges[\"all_ORF_coords\"].apply(lambda x: [i[1] for i in x])\n",
    "\n",
    "cov50_ranges[\"overlap_ORF_coords\"] = cov50_ranges.apply(lambda x: calc_overlaps(pd.DataFrame({\"Left\":x.left_ORF_coords,\"Right\":x.right_ORF_coords}),by=\"LeftRight\")[0][\"Coords\"].tolist(), axis=1)\n",
    "\n",
    "cov50_ranges[\"qranges\"] = cov50_ranges[\"qrange\"].apply(lambda x: [[[x[::2], x[1::2]][0][i], [x[::2], x[1::2]][1][i]] for i in range(len([x[::2], x[1::2]][0]))])\n",
    "\n",
    "cov50_ranges[\"in_ORF_matrix\"] = cov50_ranges.apply(lambda x: [bool(list(pd.RangeIndex(i[0],i[1]).intersection(pd.RangeIndex(b[0],b[1])))) for i in x.qranges for b in x.overlap_ORF_coords], axis=1)\n",
    "cov50_ranges[\"in_ORF_matrix_largest\"] = cov50_ranges.apply(lambda x: [bool(list(pd.RangeIndex(i[0],i[1]).intersection(pd.RangeIndex(b[0],b[1])))) for i in x.qranges for b in [x.ORF_coords]], axis=1)\n",
    "\n",
    "cov50_ranges[\"qranges_in_ORF\"] = cov50_ranges.apply(lambda x: [any(z) for z in [[bool(list(pd.RangeIndex(i[0],i[1]).intersection(pd.RangeIndex(b[0],b[1])))) for i in x.qranges for b in x.overlap_ORF_coords][v:v+len(x.overlap_ORF_coords)] for v in range(0,len(x.in_ORF_matrix),len(x.overlap_ORF_coords))]], axis=1)\n",
    "cov50_ranges[\"qranges_in_largest_ORF\"] = cov50_ranges.apply(lambda x: [any(z) for z in [[bool(list(pd.RangeIndex(i[0],i[1]).intersection(pd.RangeIndex(b[0],b[1])))) for i in x.qranges for b in [x.ORF_coords]][v:v+len([x.ORF_coords])] for v in range(0,len(x.in_ORF_matrix_largest),len([x.ORF_coords]))]], axis=1)\n",
    "\n",
    "cov50_ranges[\"qstarts_sorted_values\"] = cov50_ranges[\"qstarts\"].apply(lambda x: sort_with_order(x, make_equal=False)[0])\n",
    "cov50_ranges[\"qstarts_sorted_order\"] = cov50_ranges[\"qstarts\"].apply(lambda x: sort_with_order(x, make_equal=False)[1])\n",
    "cov50_ranges[\"qstarts_sorted_order_sequential\"] = cov50_ranges[\"qstarts\"].apply(lambda x: sort_with_order(x, make_sequential=True)[1])#true order of duplicate is lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c5c8ac-aff7-486f-8fbb-2a579a85c1ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#initialise plotting transcript arrangement\n",
    "cov50_ranges_ = cov50_ranges.sort_values(by=\"start\").reset_index(drop=True)\n",
    "cov50_ranges_[\"TrOverlaps\"] = \"\"\n",
    "cov50_ranges_[\"TrSeparate\"] = \"\"\n",
    "\n",
    "for ii, ri in cov50_ranges_.iterrows():\n",
    "    overlaps_with = []\n",
    "    start = cov50_ranges_[\"start\"][ii]\n",
    "    end = cov50_ranges_[\"end\"][ii]\n",
    "    for ij, rj in cov50_ranges_.iterrows():\n",
    "        if (end > cov50_ranges_[\"start\"][ij] and end > cov50_ranges_[\"end\"][ij] and start > cov50_ranges_[\"start\"][ij] and start > cov50_ranges_[\"end\"][ij]) or (end < cov50_ranges_[\"start\"][ij] and end < cov50_ranges_[\"end\"][ij] and start < cov50_ranges_[\"start\"][ij] and start < cov50_ranges_[\"end\"][ij]):\n",
    "            overlaps_with.append(cov50_ranges_[\"qseqid\"][ij])\n",
    "    overlaps_with_ = [x for x in cov50_ranges_[\"qseqid\"].tolist() if x not in overlaps_with]\n",
    "    overlaps_with_ = [x for x in overlaps_with_ if x not in [cov50_ranges_[\"qseqid\"][ii]]]\n",
    "    cov50_ranges_[\"TrOverlaps\"][ii] = overlaps_with_\n",
    "    cov50_ranges_[\"TrSeparate\"][ii] = overlaps_with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54d9e620-90e7-4302-8856-b409fc8336a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov50_ranges_[\"OverlapNum\"] = cov50_ranges_[\"TrOverlaps\"].apply(lambda x: len(x))\n",
    "cov50_ranges_[\"SeparateNum\"] = cov50_ranges_[\"TrSeparate\"].apply(lambda x: len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba1e96b-4376-4ed1-a3e5-a684e12e8697",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov50_ranges_bu = cov50_ranges_.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "241fe96d-cf44-4534-b834-5fda5ccf004b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov50_ranges_ =  cov50_ranges_bu.copy()\n",
    "cov50_ranges_[\"NumExon\"] = cov50_ranges_[\"srange\"].apply(lambda x: len(x)/2)\n",
    "cov50_ranges_[\"index\"] = cov50_ranges_.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36852a8a-e47f-4058-94d0-6a3922761e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "BATCH_SIZE = 40 #larger batches take exponentially longer to process, but produce more compact plot\n",
    "lrRNAseq_plot_df = pd.DataFrame()\n",
    "cov50_ranges_bu = pd.DataFrame()\n",
    "\n",
    "ranges = []\n",
    "for iii in range(0,cov50.drop_duplicates(\"qseqid\").shape[0],BATCH_SIZE):\n",
    "    ranges.append(iii)\n",
    "ranges.append(cov50.drop_duplicates(\"qseqid\").shape[0])\n",
    "\n",
    "for iii in range(len(ranges)-1):\n",
    "    print(iii)\n",
    "    cov50_ranges_ = cov50_ranges.sort_values(by=\"start\").reset_index(drop=True)[ranges[iii]:ranges[iii+1]]\n",
    "    cov50_ranges_[\"TrOverlaps\"] = \"\"\n",
    "    cov50_ranges_[\"TrSeparate\"] = \"\"\n",
    "    \n",
    "    for ii, ri in cov50_ranges_.iterrows():\n",
    "        overlaps_with = []\n",
    "        start = cov50_ranges_[\"start\"][ii]\n",
    "        end = cov50_ranges_[\"end\"][ii]\n",
    "        for ij, rj in cov50_ranges_.iterrows():\n",
    "            if (end > cov50_ranges_[\"start\"][ij] and end > cov50_ranges_[\"end\"][ij] and start > cov50_ranges_[\"start\"][ij] and start > cov50_ranges_[\"end\"][ij]) or (end < cov50_ranges_[\"start\"][ij] and end < cov50_ranges_[\"end\"][ij] and start < cov50_ranges_[\"start\"][ij] and start < cov50_ranges_[\"end\"][ij]):\n",
    "                overlaps_with.append(cov50_ranges_[\"qseqid\"][ij])\n",
    "        overlaps_with_ = [x for x in cov50_ranges_[\"qseqid\"].tolist() if x not in overlaps_with]\n",
    "        #overlaps_with_ = list(set(cov50_ranges_[\"qseqid\"].tolist())-set(overlaps_with))\n",
    "        overlaps_with_ = [x for x in overlaps_with_ if x not in [cov50_ranges_[\"qseqid\"][ii]]]\n",
    "        cov50_ranges_[\"TrOverlaps\"][ii] = overlaps_with_\n",
    "        cov50_ranges_[\"TrSeparate\"][ii] = overlaps_with\n",
    "    \n",
    "    cov50_ranges_[\"OverlapNum\"] = cov50_ranges_[\"TrOverlaps\"].apply(lambda x: len(x))\n",
    "    cov50_ranges_[\"SeparateNum\"] = cov50_ranges_[\"TrSeparate\"].apply(lambda x: len(x))\n",
    "    \n",
    "    if iii==0:\n",
    "        cov50_ranges_bu = cov50_ranges_.copy()\n",
    "    else:\n",
    "        cov50_ranges_bu = pd.concat([cov50_ranges_bu, cov50_ranges_])\n",
    "        cov50_ranges_bu = cov50_ranges_bu.reset_index(drop=True)\n",
    "    \n",
    "    cov50_ranges_[\"NumExon\"] = cov50_ranges_[\"srange\"].apply(lambda x: len(x)/2)\n",
    "    cov50_ranges_[\"index\"] = cov50_ranges_.index\n",
    "    \n",
    "    from operator import itemgetter\n",
    "    \n",
    "    def visit(TpM, chain, ii, chains,stop,chain_bu):\n",
    "        for z in range(len(TpM)):        \n",
    "            if stop == 0 and TpM[ii][z] == 1 and max(chain[:])<z:\n",
    "                chain_bu = chain[:]\n",
    "                chain.append(z)\n",
    "                if sum(TpM[z]) > 0:\n",
    "                    visit(TpM, chain, z, chains,stop,chain_bu)\n",
    "                else:\n",
    "                    chains.append(chain[:])\n",
    "                    chain=chain_bu[:]\n",
    "    \n",
    "    \n",
    "    #cov50_ranges_.sort_values(by=\"TrLen\", ascending=False, inplace=True)\n",
    "    cov50_ranges_.sort_values(by=\"TrLen\", ascending=False)\n",
    "    \n",
    "    mut_cov50_ranges = cov50_ranges_.copy()\n",
    "    TrsLeft = mut_cov50_ranges.index.tolist()\n",
    "    \n",
    "    final_coord_list = []\n",
    "    final_coord_members = []\n",
    "    final_coord_groups = []\n",
    "    final_chain_members = []\n",
    "    final_chain_ranges = []\n",
    "    final_group_lengths = []\n",
    "    saveTpM_list = []\n",
    "    for idx, r in mut_cov50_ranges.iterrows():\n",
    "        if idx in TrsLeft:\n",
    "            TrPartners = mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"TrSeparate\"].values[0]\n",
    "            #print(len(TrPartners))\n",
    "            if len(TrPartners) > 0:\n",
    "                TpM = [[0]*len(TrPartners) for _ in range(len(TrPartners))]\n",
    "                for trp in TrPartners:\n",
    "                    #print(trp)\n",
    "                    TpPartners = mut_cov50_ranges[mut_cov50_ranges[\"qseqid\"]==trp][\"TrSeparate\"].values[0]\n",
    "                    TrTpPartners = [x for x in TrPartners if x in TpPartners]\n",
    "                    # print(mut_cov50_ranges[\"qseqid\"][idx])\n",
    "                    # print(\"idx\", idx)\n",
    "                    # print(TrPartners)\n",
    "                    # print(TpPartners)\n",
    "                    # print(len(TpPartners))\n",
    "                    # print(TrTpPartners)\n",
    "                    # print(len(TrTpPartners))\n",
    "                    for trtpp in TrTpPartners:\n",
    "                        if TrPartners.index(trp)<TrPartners.index(trtpp):\n",
    "                            TpM[TrPartners.index(trp)][TrPartners.index(trtpp)] = 1\n",
    "    \n",
    "                if len(TpM)>1:\n",
    "                    saveTpM_list.append(TpM)\n",
    "                if sum([x for i in TpM for x in i]) == 0:\n",
    "                    \n",
    "                    total_lengths = []\n",
    "                    \n",
    "                    for tr in TrPartners:\n",
    "                        total_length = mut_cov50_ranges[mut_cov50_ranges[\"qseqid\"] == tr][\"TrLen\"].tolist()[0]\n",
    "                        total_lengths.append(total_length)\n",
    "                    \n",
    "                    index_max = max(range(len(total_lengths)), key=total_lengths.__getitem__)\n",
    "                    \n",
    "                    best_trs_df = pd.concat([mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx], mut_cov50_ranges[mut_cov50_ranges[\"qseqid\"] == TrPartners[index_max]]])\n",
    "                    best_trs_df = best_trs_df.sort_values(by=\"start\")\n",
    "                    coord_list = best_trs_df[\"srange\"].tolist()\n",
    "                    coord_list_ = [x for l in coord_list for x in l]\n",
    "                    coord_members = len(coord_list_)/2\n",
    "                    coord_groups = best_trs_df[\"NumExon\"].tolist()\n",
    "                    \n",
    "                    final_coord_list.append(coord_list_)\n",
    "                    final_coord_members.append(coord_members)\n",
    "                    final_chain_members.append(best_trs_df[\"qseqid\"].tolist())\n",
    "                    final_coord_groups.append(coord_groups)\n",
    "                    final_chain_ranges.append(best_trs_df[\"TrRange\"].tolist())\n",
    "                    final_group_lengths.append(best_trs_df[\"TrLen\"].sum())\n",
    "                    \n",
    "                    trs_to_remove = best_trs_df.index.tolist()\n",
    "                    TrsLeft = [x for x in TrsLeft if x not in trs_to_remove]\n",
    "                    \n",
    "                    cov50_ranges_.drop(trs_to_remove, inplace=True)\n",
    "                    \n",
    "                    for ii, ri in cov50_ranges_.iterrows():\n",
    "                        overlaps_with = []\n",
    "                        start = cov50_ranges_[\"start\"][ii]\n",
    "                        end = cov50_ranges_[\"end\"][ii]\n",
    "                        for ij, rj in cov50_ranges_.iterrows():\n",
    "                            if (end > cov50_ranges_[\"start\"][ij] and end > cov50_ranges_[\"end\"][ij] and start > cov50_ranges_[\"start\"][ij] and start > cov50_ranges_[\"end\"][ij]) or (end < cov50_ranges_[\"start\"][ij] and end < cov50_ranges_[\"end\"][ij] and start < cov50_ranges_[\"start\"][ij] and start < cov50_ranges_[\"end\"][ij]):\n",
    "                                overlaps_with.append(cov50_ranges_[\"qseqid\"][ij])\n",
    "                        overlaps_with_ = [x for x in cov50_ranges_[\"qseqid\"].tolist() if x not in overlaps_with]\n",
    "                        #overlaps_with_ = list(set(cov50_ranges_[\"qseqid\"].tolist())-set(overlaps_with))\n",
    "                        overlaps_with_ = [x for x in overlaps_with_ if x not in [cov50_ranges_[\"qseqid\"][ii]]]\n",
    "                        cov50_ranges_[\"TrOverlaps\"][ii] = overlaps_with_\n",
    "                        cov50_ranges_[\"TrSeparate\"][ii] = overlaps_with\n",
    "                    \n",
    "                    cov50_ranges_[\"OverlapNum\"] = cov50_ranges_[\"TrOverlaps\"].apply(lambda x: len(x))\n",
    "                    cov50_ranges_[\"SeparateNum\"] = cov50_ranges_[\"TrSeparate\"].apply(lambda x: len(x))\n",
    "                \n",
    "                    mut_cov50_ranges = cov50_ranges_.copy()\n",
    "                    \n",
    "                else:\n",
    "                    chains = []\n",
    "                    stop = 0\n",
    "                    for i in range(len(TpM)):\n",
    "                        for j in range(len(TpM)):\n",
    "                            if TpM[i][j] == 1 and stop==0:\n",
    "                                chain = [i,j] #0,7\n",
    "                                chain_bkup = chain[:]\n",
    "                                chains.append(chain[:])\n",
    "                                if sum(TpM[j]) > 0:\n",
    "                                    visit(TpM, chain, j, chains, stop, chain_bkup)\n",
    "                    \n",
    "                    correction = []\n",
    "                    for elem in chains:\n",
    "                        if sum(TpM[elem[-1]])==0 and sum(TpM[elem[-2]])==0:\n",
    "                            correction.append(elem)\n",
    "                    \n",
    "                    chains = [x for x in chains if x not in correction]\n",
    "                    new_chains = []\n",
    "                    for elem in chains:\n",
    "                        if elem not in new_chains:\n",
    "                            new_chains.append(elem)\n",
    "                    chains = new_chains\n",
    "                   # print(chains)\n",
    "                    total_lengths = []\n",
    "                    \n",
    "                    for chain in chains:\n",
    "                        total_length = 0\n",
    "                        for i in chain:\n",
    "                            total_length += mut_cov50_ranges[mut_cov50_ranges[\"qseqid\"] == TrPartners[i]][\"TrLen\"].tolist()[0]\n",
    "                        total_lengths.append(total_length)\n",
    "                    #print(total_lengths)\n",
    "                    index_max = max(range(len(total_lengths)), key=total_lengths.__getitem__)\n",
    "                \n",
    "                    best_chain = chains[index_max]\n",
    "                \n",
    "                    best_trs = itemgetter(*best_chain)(TrPartners)\n",
    "                    #print(\"besttrs1\", best_trs)\n",
    "                    #print(\"idx\", idx)\n",
    "                    best_trs_df = pd.concat([mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx], mut_cov50_ranges[mut_cov50_ranges[\"qseqid\"].isin(best_trs)]])\n",
    "                    #print(\"besttrs2\", best_trs_df)\n",
    "                    best_trs_df = best_trs_df.sort_values(by=\"start\")\n",
    "                    coord_list = best_trs_df[\"srange\"].tolist()\n",
    "                    coord_list_ = [x for l in coord_list for x in l]\n",
    "                    coord_members = len(coord_list_)/2\n",
    "                    coord_groups = best_trs_df[\"NumExon\"].tolist()\n",
    "                    \n",
    "                    final_coord_list.append(coord_list_)\n",
    "                    final_coord_members.append(coord_members)\n",
    "                    final_chain_members.append(best_trs_df[\"qseqid\"].tolist())\n",
    "                    final_coord_groups.append(coord_groups)\n",
    "                    final_chain_ranges.append(best_trs_df[\"TrRange\"].tolist())\n",
    "                    final_group_lengths.append(best_trs_df[\"TrLen\"].sum())\n",
    "                \n",
    "                    trs_to_remove = best_trs_df.index.tolist()\n",
    "                    TrsLeft = [x for x in TrsLeft if x not in trs_to_remove]\n",
    "                    \n",
    "                    cov50_ranges_.drop(trs_to_remove, inplace=True)\n",
    "                    \n",
    "                    for ii, ri in cov50_ranges_.iterrows():\n",
    "                        overlaps_with = []\n",
    "                        start = cov50_ranges_[\"start\"][ii]\n",
    "                        end = cov50_ranges_[\"end\"][ii]\n",
    "                        for ij, rj in cov50_ranges_.iterrows():\n",
    "                            if (end > cov50_ranges_[\"start\"][ij] and end > cov50_ranges_[\"end\"][ij] and start > cov50_ranges_[\"start\"][ij] and start > cov50_ranges_[\"end\"][ij]) or (end < cov50_ranges_[\"start\"][ij] and end < cov50_ranges_[\"end\"][ij] and start < cov50_ranges_[\"start\"][ij] and start < cov50_ranges_[\"end\"][ij]):\n",
    "                                overlaps_with.append(cov50_ranges_[\"qseqid\"][ij])\n",
    "                        overlaps_with_ = [x for x in cov50_ranges_[\"qseqid\"].tolist() if x not in overlaps_with]\n",
    "                        #overlaps_with_ = list(set(cov50_ranges_[\"qseqid\"].tolist())-set(overlaps_with))\n",
    "                        overlaps_with_ = [x for x in overlaps_with_ if x not in [cov50_ranges_[\"qseqid\"][ii]]]\n",
    "                        cov50_ranges_[\"TrOverlaps\"][ii] = overlaps_with_\n",
    "                        cov50_ranges_[\"TrSeparate\"][ii] = overlaps_with\n",
    "                    \n",
    "                    cov50_ranges_[\"OverlapNum\"] = cov50_ranges_[\"TrOverlaps\"].apply(lambda x: len(x))\n",
    "                    cov50_ranges_[\"SeparateNum\"] = cov50_ranges_[\"TrSeparate\"].apply(lambda x: len(x))\n",
    "                \n",
    "                    mut_cov50_ranges = cov50_ranges_.copy()\n",
    "            else:\n",
    "                coord_list_ = mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"srange\"].values[0]\n",
    "                coord_members = len(coord_list_)/2            \n",
    "                coord_groups = mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"NumExon\"].values[0]\n",
    "                \n",
    "                final_coord_list.append(coord_list_)\n",
    "                final_coord_members.append(coord_members)\n",
    "                final_chain_members.append(mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"qseqid\"].values[0])\n",
    "                final_coord_groups.append(coord_groups)\n",
    "                final_chain_ranges.append(mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"TrRange\"].values[0])\n",
    "                final_group_lengths.append(mut_cov50_ranges[mut_cov50_ranges[\"index\"] == idx][\"TrLen\"].values[0])\n",
    "                \n",
    "                trs_to_remove = [idx]\n",
    "                TrsLeft = [x for x in TrsLeft if x not in trs_to_remove]\n",
    "                \n",
    "                cov50_ranges_.drop(trs_to_remove, inplace=True)\n",
    "                \n",
    "                for ii, ri in cov50_ranges_.iterrows():\n",
    "                    overlaps_with = []\n",
    "                    start = cov50_ranges_[\"start\"][ii]\n",
    "                    end = cov50_ranges_[\"end\"][ii]\n",
    "                    for ij, rj in cov50_ranges_.iterrows():\n",
    "                        if (end > cov50_ranges_[\"start\"][ij] and end > cov50_ranges_[\"end\"][ij] and start > cov50_ranges_[\"start\"][ij] and start > cov50_ranges_[\"end\"][ij]) or (end < cov50_ranges_[\"start\"][ij] and end < cov50_ranges_[\"end\"][ij] and start < cov50_ranges_[\"start\"][ij] and start < cov50_ranges_[\"end\"][ij]):\n",
    "                            overlaps_with.append(cov50_ranges_[\"qseqid\"][ij])\n",
    "                    overlaps_with_ = [x for x in cov50_ranges_[\"qseqid\"].tolist() if x not in overlaps_with]\n",
    "                    #overlaps_with_ = list(set(cov50_ranges_[\"qseqid\"].tolist())-set(overlaps_with))\n",
    "                    overlaps_with_ = [x for x in overlaps_with_ if x not in [cov50_ranges_[\"qseqid\"][ii]]]\n",
    "                    cov50_ranges_[\"TrOverlaps\"][ii] = overlaps_with_\n",
    "                    cov50_ranges_[\"TrSeparate\"][ii] = overlaps_with\n",
    "                \n",
    "                cov50_ranges_[\"OverlapNum\"] = cov50_ranges_[\"TrOverlaps\"].apply(lambda x: len(x))\n",
    "                cov50_ranges_[\"SeparateNum\"] = cov50_ranges_[\"TrSeparate\"].apply(lambda x: len(x))\n",
    "            \n",
    "                mut_cov50_ranges = cov50_ranges_.copy()\n",
    "    \n",
    "    for idx, row in mut_cov50_ranges.iterrows():\n",
    "        coord_list_ = mut_cov50_ranges[\"srange\"][idx]\n",
    "        coord_members = len(coord_list_)/2            \n",
    "        coord_groups = mut_cov50_ranges[\"NumExon\"][idx]\n",
    "        \n",
    "        final_coord_list.append(coord_list_)\n",
    "        final_coord_members.append(coord_members)\n",
    "        final_chain_members.append(mut_cov50_ranges[\"qseqid\"][idx])\n",
    "        final_coord_groups.append(coord_groups)\n",
    "        final_chain_ranges.append(mut_cov50_ranges[\"TrRange\"][idx])\n",
    "        final_group_lengths.append(mut_cov50_ranges[\"TrLen\"][idx])\n",
    "    \n",
    "    lrRNAseq_plot_df_pre = pd.DataFrame()\n",
    "    lrRNAseq_plot_df_pre[\"coords\"] = final_coord_list\n",
    "    lrRNAseq_plot_df_pre[\"total_exons\"] = final_coord_members\n",
    "    lrRNAseq_plot_df_pre[\"TrNames\"] = final_chain_members\n",
    "    lrRNAseq_plot_df_pre[\"exons_per_tr\"] = final_coord_groups\n",
    "    lrRNAseq_plot_df_pre[\"range_per_tr\"] = final_chain_ranges\n",
    "    lrRNAseq_plot_df_pre[\"Total_aln_len\"] = final_group_lengths\n",
    "    lrRNAseq_plot_df = pd.concat([lrRNAseq_plot_df, lrRNAseq_plot_df_pre])\n",
    "    lrRNAseq_plot_df = lrRNAseq_plot_df.sort_values(\"Total_aln_len\",ascending=False)\n",
    "    lrRNAseq_plot_df = lrRNAseq_plot_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b888b40-bc10-4dac-8bda-3c42a50bafec",
   "metadata": {},
   "outputs": [],
   "source": [
    "cov50_ranges_ =  cov50_ranges_bu.copy()\n",
    "cov50_ranges_[\"NumExon\"] = cov50_ranges_[\"srange\"].apply(lambda x: len(x)/2)\n",
    "cov50_ranges_[\"index\"] = cov50_ranges_.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7021e760-51c8-40d0-86e6-b711c8b208ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine transcript plot dataframe with the filtered BLAST dataframe\n",
    "lrRNAseq_plot_df[\"ORF_coords\"] = \"\"\n",
    "lrRNAseq_plot_df[\"all_ORF_coords\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qranges_in_largest_ORF\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qranges_in_ORF\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_values\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_order\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"] = \"\"\n",
    "lrRNAseq_plot_df[\"qstarts\"] = \"\"\n",
    "lrRNAseq_plot_df[\"index\"] = lrRNAseq_plot_df.index\n",
    "\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    qstarts_sorted_order_sequential_list = []\n",
    "    qstarts_sorted_order_list = []\n",
    "    qstarts_sorted_values_list = []\n",
    "    qranges_in_largest_ORF_list = []\n",
    "    qranges_in_ORF_list = []\n",
    "    all_ORF_coords_list = []\n",
    "    ORF_coords_list = []\n",
    "    qstarts_list = []\n",
    "    if type(lrRNAseq_plot_df[\"TrNames\"][idx])==str:\n",
    "        id = lrRNAseq_plot_df[\"TrNames\"][idx]\n",
    "        qstarts_sorted_order_sequential_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_order_sequential\"].tolist()[0])\n",
    "        qstarts_sorted_order_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_order\"].tolist()[0])\n",
    "        qstarts_sorted_values_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_values\"].tolist()[0])\n",
    "        qranges_in_largest_ORF_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qranges_in_largest_ORF\"].tolist()[0])\n",
    "        qranges_in_ORF_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qranges_in_ORF\"].tolist()[0])\n",
    "        all_ORF_coords_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"all_ORF_coords\"].tolist()[0])\n",
    "        ORF_coords_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"ORF_coords\"].tolist()[0])\n",
    "        qstarts_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts\"].tolist()[0])\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"ORF_coords\"] = str(ORF_coords_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"all_ORF_coords\"] = str(all_ORF_coords_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qranges_in_largest_ORF\"] = str(qranges_in_largest_ORF_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qranges_in_ORF\"] = str(qranges_in_ORF_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_values\"] = str(qstarts_sorted_values_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_order\"] = str(qstarts_sorted_order_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_order_sequential\"] = str(qstarts_sorted_order_sequential_list)\n",
    "        lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts\"] = str(qstarts_list)\n",
    "    else:    \n",
    "        for id in lrRNAseq_plot_df[\"TrNames\"][idx]:\n",
    "            qstarts_sorted_order_sequential_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_order_sequential\"].tolist()[0])\n",
    "            qstarts_sorted_order_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_order\"].tolist()[0])\n",
    "            qstarts_sorted_values_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts_sorted_values\"].tolist()[0])\n",
    "            qranges_in_largest_ORF_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qranges_in_largest_ORF\"].tolist()[0])\n",
    "            qranges_in_ORF_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qranges_in_ORF\"].tolist()[0])\n",
    "            all_ORF_coords_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"all_ORF_coords\"].tolist()[0])\n",
    "            ORF_coords_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"ORF_coords\"].tolist()[0])\n",
    "            qstarts_list.append(cov50_ranges_[cov50_ranges_[\"qseqid\"]==id][\"qstarts\"].tolist()[0])\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"ORF_coords\"] = str(ORF_coords_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"all_ORF_coords\"] = str(all_ORF_coords_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qranges_in_largest_ORF\"] = str(qranges_in_largest_ORF_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qranges_in_ORF\"] = str(qranges_in_ORF_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_values\"] = str(qstarts_sorted_values_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_order\"] = str(qstarts_sorted_order_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts_sorted_order_sequential\"] = str(qstarts_sorted_order_sequential_list)\n",
    "            lrRNAseq_plot_df.loc[lrRNAseq_plot_df[\"index\"]== idx, \"qstarts\"] = str(qstarts_list)\n",
    "\n",
    "lrRNAseq_plot_df[\"ORF_coords\"] = lrRNAseq_plot_df[\"ORF_coords\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"all_ORF_coords\"] = lrRNAseq_plot_df[\"all_ORF_coords\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qranges_in_largest_ORF\"] = lrRNAseq_plot_df[\"qranges_in_largest_ORF\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qranges_in_ORF\"] = lrRNAseq_plot_df[\"qranges_in_ORF\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_values\"] = lrRNAseq_plot_df[\"qstarts_sorted_values\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_order\"] = lrRNAseq_plot_df[\"qstarts_sorted_order\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"] = lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"].apply(lambda x: eval(x))\n",
    "lrRNAseq_plot_df[\"qstarts\"] = lrRNAseq_plot_df[\"qstarts\"].apply(lambda x: eval(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9abf9f0f-3df9-4f08-a413-761bbeab9b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotting requirements\n",
    "lrRNAseq_plot_df = lrRNAseq_plot_df.sort_values(by=\"Total_aln_len\", ascending=False)\n",
    "lrRNAseq_plot_df[\"coverage\"] = lrRNAseq_plot_df[\"total_exons\"].apply(lambda x: [0]+[0,1,1,0]*int(x)+[0] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e27088e2-595e-4f3f-a54a-b0605747d969",
   "metadata": {},
   "outputs": [],
   "source": [
    "#further processing of plotting coordinates\n",
    "lrRNAseq_plot_df[\"coords2\"] = \"\"\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    coords = lrRNAseq_plot_df[\"coords\"][idx]\n",
    "    coords1 = coords[::2]\n",
    "    coords1_add = [x - 1 for x in coords1]\n",
    "    coords2 = coords[1::2]\n",
    "    coords2_add = [x + 1 for x in coords2]\n",
    "    lrRNAseq_plot_df[\"coords2\"][idx] = lrRNAseq_plot_df[\"coords\"][idx] + coords1_add + coords2_add \n",
    "lrRNAseq_plot_df[\"coords2\"] = lrRNAseq_plot_df[\"coords2\"].apply(lambda x: [0]+x+[len(my_seq)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e20f947-9881-460c-8e9c-1f58f675cc7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plotting coordinate statistics\n",
    "lrRNAseq_plot_df[\"min\"] = lrRNAseq_plot_df[\"coords\"].apply(lambda x: min(x))\n",
    "lrRNAseq_plot_df[\"max\"] = lrRNAseq_plot_df[\"coords\"].apply(lambda x: max(x))\n",
    "lrRNAseq_plot_df[\"span\"] = lrRNAseq_plot_df[\"max\"] - lrRNAseq_plot_df[\"min\"]\n",
    "lrRNAseq_plot_df.sort_values(by=\"span\", ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4981f2b-f9a6-4cba-9289-b60d6180c3d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#translate order and get coord ranges\n",
    "lrRNAseq_plot_df[\"qstarts_sorted_order_translated\"] = lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"].apply(lambda x: [translate_order(i) for i in x])\n",
    "lrRNAseq_plot_df[\"coord_ranges\"] = lrRNAseq_plot_df[\"coords\"].apply(lambda x: [[[x[::2],x[1::2]][0][i],[x[::2],x[1::2]][1][i]]for i in range(len([x[::2],x[1::2]][0]))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9c3f8d-3275-4255-95bc-3c0570155dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create order dictionary for each transcript\n",
    "dict_list_outer = []\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    dict_list = []\n",
    "    for i in range(len(lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"][idx])):        \n",
    "        d = {}\n",
    "        for j in range(len(lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"][idx][i])):\n",
    "            d[lrRNAseq_plot_df[\"qstarts_sorted_values\"][idx][i][j] ] = lrRNAseq_plot_df[\"qstarts_sorted_order_sequential\"][idx][i][j]\n",
    "        dict_list.append(d)\n",
    "    dict_list_outer.append(dict_list)\n",
    "lrRNAseq_plot_df[\"order_dict\"] = dict_list_outer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f411b89-a20b-4236-a13f-2b1b4bba7de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use order dictionary to check order of alignment regions in transcripts\n",
    "ordered_list_outer = []\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    ordered_list = []\n",
    "    for j in range(len(lrRNAseq_plot_df[\"order_dict\"][idx])):\n",
    "        \n",
    "        o_dict = lrRNAseq_plot_df[\"order_dict\"][idx][j]\n",
    "        qstarts = lrRNAseq_plot_df[\"qstarts\"][idx][j]\n",
    "        ordered = []\n",
    "        \n",
    "        for i in range(len(qstarts)):\n",
    "            ordered.append(o_dict[qstarts[i]])\n",
    "        ordered_list.append(ordered)\n",
    "    ordered_list_outer.append(ordered_list)\n",
    "lrRNAseq_plot_df[\"exon_synteny\"] = ordered_list_outer\n",
    "lrRNAseq_plot_df[\"centre_coords\"] = lrRNAseq_plot_df[\"coord_ranges\"].apply(lambda x: [int(i[0]+((i[1]-i[0])/2)) for i in x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7077766b-db11-4c35-af33-7d4848b26505",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ordered exon indexes and correction for closely bundled centre coordinates for better visibility of the numbers in the plot\n",
    "lrRNAseq_plot_df[\"syntenous_indexes\"] = lrRNAseq_plot_df[\"exon_synteny\"].apply(lambda x: [syntenous_indexes(i) for i in x])\n",
    "lrRNAseq_plot_df[\"centre_coords_corrected\"] = lrRNAseq_plot_df[\"centre_coords\"].apply(lambda x: space_elements(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61369be6-47d3-4ad2-864b-6c3810408e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#required for plotting multi-transcript syntentic regions\n",
    "#convert exons_per_tr to lists\n",
    "exons_per_tr_list = []\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    ex_per_tr = lrRNAseq_plot_df[\"exons_per_tr\"][idx]\n",
    "    if type(ex_per_tr) == float:\n",
    "        exons_per_tr_list.append([int(ex_per_tr)])\n",
    "    else:\n",
    "        exons_per_tr_list.append(ex_per_tr)\n",
    "lrRNAseq_plot_df[\"exons_per_tr\"] = exons_per_tr_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b5c767e-bf08-4e1a-947b-c88358ebb724",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make syntenous_indexes lists have even number of members\n",
    "new_xs = []\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    x = lrRNAseq_plot_df[\"syntenous_indexes\"][idx].copy()\n",
    "# x = [[0,3,0],[0],[0,4,5,7]]\n",
    "    new_x = []\n",
    "    for i in x:\n",
    "        if len(i)%2!=0:\n",
    "            i.append(i[len(i)-1])\n",
    "        new_x.append(i)\n",
    "    new_xs.append(new_x)\n",
    "lrRNAseq_plot_df[\"syntenous_indexes_even\"] = new_xs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "397b7244-f1cf-4cb4-99b8-21250f4eb00d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#increment subsequent transcript lists by number of exons in previous transcripts\n",
    "syn_idxs_corr_list = []\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "    syn_idxs = lrRNAseq_plot_df[\"syntenous_indexes_even\"][idx].copy()\n",
    "    # syn_idxs = [[0, 1], [0, 5, 6, 7, 8, 14, 15, 16, 22],[0,3]]\n",
    "    ex_per_tr = lrRNAseq_plot_df[\"exons_per_tr\"][idx].copy()\n",
    "    # ex_per_tr = [2.0, 23.0, 4]\n",
    "    num_trs = len(syn_idxs)\n",
    "    increment = 0\n",
    "    syn_idxs_corrected = []\n",
    "    syn_idxs_corrected.append(syn_idxs[0])\n",
    "    \n",
    "    for i in range(1,num_trs):\n",
    "        increment += ex_per_tr[i-1]\n",
    "        syn_idxs_corrected.append((np.array(syn_idxs[i])+increment).tolist())\n",
    "    syn_idxs_corr_list.append(syn_idxs_corrected)\n",
    "lrRNAseq_plot_df[\"syn_idxs_corrected\"] = syn_idxs_corr_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48c641d2-0f4c-415c-83a7-74276a5722e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#finalising information\n",
    "lrRNAseq_plot_df[\"qranges_in_ORF_unwrapped\"] = lrRNAseq_plot_df[\"qranges_in_ORF\"].apply(lambda x: [i for i in extract_nested_values(x)] )\n",
    "lrRNAseq_plot_df[\"qranges_in_largest_ORF_unwrapped\"] = lrRNAseq_plot_df[\"qranges_in_largest_ORF\"].apply(lambda x: [i for i in extract_nested_values(x)] )\n",
    "lrRNAseq_plot_df[\"exon_synteny_unwrapped\"] = lrRNAseq_plot_df[\"exon_synteny\"].apply(lambda x: [i for i in extract_nested_values(x)] )\n",
    "lrRNAseq_plot_df[\"syntenous_indexes_unwrapped\"] = lrRNAseq_plot_df[\"syn_idxs_corrected\"].apply(lambda x: [i for i in extract_nested_values(x)] )\n",
    "lrRNAseq_plot_df[\"syntenous_indexes_int_unwrapped\"]  = lrRNAseq_plot_df[\"syntenous_indexes_unwrapped\"].apply(lambda x: [int(i) for i in x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b6727c1-7488-4c55-85e0-9a1d4277d678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot results\n",
    "seq_length = seq_len\n",
    "matplotlib.rcParams[\"agg.path.chunksize\"] = 10000\n",
    "plt.rcParams['agg.path.chunksize'] = 10000\n",
    "figsize_param_w = min(int(seq_length/1100), 50)\n",
    "figsize_param_h = int(lrRNAseq_plot_df.shape[0]/2.5)\n",
    "plt.rcParams['figure.figsize'] = [figsize_param_w, figsize_param_h]\n",
    "plt.rcParams['figure.dpi'] = 140\n",
    "\n",
    "nrows = lrRNAseq_plot_df.shape[0]\n",
    "fig, axes = plt.subplots(nrows=nrows, ncols=1, sharex=False)\n",
    "plt.subplots_adjust(top=0.9, bottom=0.1, hspace=0)\n",
    "axes[lrRNAseq_plot_df.shape[0]-1].set_xlabel(\"\")\n",
    "\n",
    "\n",
    "lrRNAseq_plot_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "for idx, row in lrRNAseq_plot_df.iterrows():\n",
    "#plot all alignment regions\n",
    "    temp_plot_df = pd.DataFrame()\n",
    "    temp_plot_df[\"coords\"] = lrRNAseq_plot_df[\"coords2\"][idx]\n",
    "    temp_plot_df.sort_values(by=\"coords\", inplace=True)\n",
    "    temp_plot_df[\"coverage\"] = lrRNAseq_plot_df[\"coverage\"][idx]\n",
    "    temp_plot_df.plot.area(x=\"coords\", y= \"coverage\", figsize=(figsize_param_w, figsize_param_h), grid=True,\n",
    "                           legend=False, ax = axes[idx], color='r', linewidth=1, xlim=[0,seq_length],zorder=1)\n",
    "#plot line if only 1 line in lane\n",
    "    temp_lines = pd.DataFrame()\n",
    "    if type(lrRNAseq_plot_df[\"range_per_tr\"][idx][0]) == int:\n",
    "        temp_lines[\"coord\"] = lrRNAseq_plot_df[\"range_per_tr\"][idx]\n",
    "        temp_lines[\"midline\"] = 0.5\n",
    "        temp_lines.plot.line(x=\"coord\", y= \"midline\", figsize=(figsize_param_w, figsize_param_h), grid=True, \n",
    "                             legend=False, ax = axes[idx], color='b', linewidth=3, xlim=[0,seq_length])\n",
    "#plot line if multiple lines in lane\n",
    "    else:\n",
    "        c = [\"b\",\"g\",\"m\",\"c\",\"w\",\"grey\",\"k\", \"silver\", \"beige\",\"teal\",\"lightskyblue\",\"r\",\"y\",\"lightgreen\",\"peru\"]\n",
    "        i=0\n",
    "        for line in lrRNAseq_plot_df[\"range_per_tr\"][idx]:\n",
    "            temp_lines[\"coord\"] = line\n",
    "            temp_lines[\"midline\"] = 0.5\n",
    "            temp_lines.plot.line(x=\"coord\", y= \"midline\", figsize=(figsize_param_w, figsize_param_h), grid=True,\n",
    "                                   legend=False, ax = axes[idx], color=c[i], linewidth=4, xlim=[0,seq_length])\n",
    "            i+=1\n",
    "#plot alignment region in ORF as a rectangle patch - border only\n",
    "    v = 0\n",
    "    coord_ranges = lrRNAseq_plot_df[\"coord_ranges\"][idx]\n",
    "    qranges_in_ORF = lrRNAseq_plot_df[\"qranges_in_ORF_unwrapped\"][idx]\n",
    "    # qranges_in_ORF = lrRNAseq_plot_df[\"qranges_in_ORF\"][idx]\n",
    "    for c in coord_ranges:\n",
    "        if qranges_in_ORF[v] == True:\n",
    "            a2 = patches.Rectangle((c[0],0),c[1]-c[0],1,color=\"gold\", fill=None,zorder=1)\n",
    "            axes[idx].add_patch(a2)\n",
    "        v+=1\n",
    "#plot alignment region in largest ORF as a rectangle patch - border and fill\n",
    "    v = 0\n",
    "    qranges_in_largest_ORF = lrRNAseq_plot_df[\"qranges_in_largest_ORF_unwrapped\"][idx]\n",
    "    for c in coord_ranges:\n",
    "        if qranges_in_largest_ORF[v] == True:\n",
    "            a1 = patches.Rectangle((c[0],0),c[1]-c[0],1,color=\"gold\",zorder=99)\n",
    "            a2 = patches.Rectangle((c[0],0),c[1]-c[0],1,color=\"gold\", fill=None,zorder=1)\n",
    "            axes[idx].add_patch(a1)\n",
    "            axes[idx].add_patch(a2)\n",
    "        v+=1\n",
    "\n",
    "#plot text of syntentic order \n",
    "    centre_coords = lrRNAseq_plot_df[\"centre_coords\"][idx]\n",
    "    syntentic_order = lrRNAseq_plot_df[\"exon_synteny_unwrapped\"][idx]\n",
    "    for i in range(len(centre_coords)):\n",
    "        fig.text(centre_coords[i], int((nrows-idx)/(nrows+1))+0.5, f\"{syntentic_order[i]}\", horizontalalignment=\"center\", transform=axes[idx].transData, verticalalignment=\"center\", fontsize=15, color=\"black\")\n",
    "\n",
    "#plot syntentic frame\n",
    "    v = 0\n",
    "    coord_ranges = lrRNAseq_plot_df[\"coord_ranges\"][idx]\n",
    "    structure = lrRNAseq_plot_df[\"syntenous_indexes_int_unwrapped\"][idx]\n",
    "    syntentic_coords = []\n",
    "    for i in range(0,len(structure)-1,2):\n",
    "        start = coord_ranges[structure[i]][0]\n",
    "        end = coord_ranges[structure[i+1]][1]\n",
    "        syntentic_coords.append([start,end])\n",
    "    for c in syntentic_coords:\n",
    "        # a4 = patches.Rectangle((c[0],0)c[1]-c[0],1,color=\"pink\",zorder=1)\n",
    "        a3 = patches.Rectangle((c[0],0),c[1]-c[0],1,color=\"pink\",zorder=0)\n",
    "        axes[idx].add_patch(a3)\n",
    "        # axes[idx].add_patch(a4)\n",
    "        v+=1\n",
    "\n",
    "for idx in range(nrows):\n",
    "    if idx != 0 and idx != nrows-1:\n",
    "        axes[idx].set_xticklabels([])\n",
    "        axes[idx].set_xlabel(\"\")\n",
    "        axes[idx].minorticks_on()\n",
    "        \n",
    "        # axes[idx].tick_params(axis='both', which='major', labelsize=30, pad=15, width=3, size=8) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        # axes[idx].tick_params(axis='both', which='minor', labelsize=30, pad=15, width=2, size=5) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='major', labelsize=30, pad=0, width=0, size=0) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='minor', labelsize=0, pad=0, width=0, size=0) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='major', labelsize=30, pad=15, width=3, size=8,labelbottom=True,labeltop=True,bottom=True,top=True) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='minor', labelsize=30, pad=15, width=2, size=5,labelbottom=True,labeltop=True,bottom=True,top=True) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].grid(which=\"major\", axis=\"x\", color = 'black', linestyle = '--', dashes=(5, 5), linewidth = 1.35, alpha=0.65) \n",
    "        axes[idx].grid(which=\"minor\", axis=\"x\", color = 'black', linestyle = '-', dashes=(4, 5), linewidth = 0.85, alpha=0.45)    \n",
    "        for axis in [\"top\", \"bottom\", \"left\", \"right\"]: \n",
    "            axes[idx].spines[axis].set_linewidth(2.5)\n",
    "        axes[0].xaxis.tick_top()\n",
    "        \n",
    "    elif idx == 0:\n",
    "        axes[idx].set_xlabel(\"\")\n",
    "        axes[idx].minorticks_on()\n",
    "        \n",
    "        # axes[idx].tick_params(axis='both', which='major', labelsize=30, pad=15, width=3, size=8) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        # axes[idx].tick_params(axis='both', which='minor', labelsize=30, pad=15, width=2, size=5) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='major', labelsize=30, pad=0, width=0, size=0) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='minor', labelsize=0, pad=0, width=0, size=0) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='major', labelsize=30, pad=15, width=3, size=8,labelbottom=0,labeltop=True,bottom=0,top=True) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='minor', labelsize=30, pad=15, width=2, size=5,labelbottom=0,labeltop=True,bottom=0,top=True) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].grid(which=\"major\", axis=\"x\", color = 'black', linestyle = '--', dashes=(5, 5), linewidth = 1.35, alpha=0.65) \n",
    "        axes[idx].grid(which=\"minor\", axis=\"x\", color = 'black', linestyle = '-', dashes=(4, 5), linewidth = 0.85, alpha=0.45)    \n",
    "        for axis in [\"top\", \"bottom\", \"left\", \"right\"]: \n",
    "            axes[idx].spines[axis].set_linewidth(2.5)\n",
    "        axes[0].xaxis.tick_top()\n",
    "\n",
    "    elif idx == nrows-1:\n",
    "        # axes[idx].tick_params(axis='both', which='major', labelsize=30, pad=15, width=3, size=8) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        # axes[idx].tick_params(axis='both', which='minor', labelsize=30, pad=15, width=2, size=5) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='major', labelsize=30, pad=0, width=0, size=0) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='y', which='minor', labelsize=0, pad=0, width=0, size=0) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='major', labelsize=30, pad=15, width=3, size=8,labelbottom=True,labeltop=0,bottom=True,top=0) #labelbottom=True,labeltop=True,bottom=True,top=True\n",
    "        axes[idx].tick_params(axis='x', which='minor', labelsize=30, pad=15, width=2, size=5,labelbottom=True,labeltop=0,bottom=True,top=0) #labeltop=True,bottom=True,top=True\n",
    "        axes[idx].grid(which=\"major\", axis=\"x\", color = 'black', linestyle = '--', dashes=(5, 5), linewidth = 1.35, alpha=0.65) \n",
    "        axes[idx].grid(which=\"minor\", axis=\"x\", color = 'black', linestyle = '-', dashes=(4, 5), linewidth = 0.85, alpha=0.45)    \n",
    "        for axis in [\"top\", \"bottom\", \"left\", \"right\"]: \n",
    "            axes[idx].spines[axis].set_linewidth(2.5)\n",
    "        axes[0].xaxis.tick_top()\n",
    "        axes[idx].set_xlabel(\"\")\n",
    "        axes[idx].minorticks_on()\n",
    "        \n",
    "    for z, label in enumerate(axes[idx].get_yticklabels()):\n",
    "        if z >= 0 and z < len(axes[idx].get_yticklabels()):\n",
    "            label.set_visible(False)\n",
    "    \n",
    "    axes[idx].set_yticklabels([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b003d2f-34b1-4cc0-9ccb-873144497e5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#the rows of plot are indexed by the index of lrRNAseq_plot_df dataframe\n",
    "#TrNames column of lrRNAseq_plot_df contains the names of transcript for the corresponding row of the plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
